### PR TITLE
Bug 567014 - Avoid newPlexusContainer(..) at startup

### PR DESCRIPTION
**Pull request by mickaelistria**
Delay creation of container (which is expensive) later in m2e lifecycle,
only when it's necessary.

Change-Id: I00f3bc645e0b962bf5c914293b001869ae73e928
Signed-off-by: Mickael Istria <mistria@redhat.com>

### CODE REVIEW

File: org.eclipse.m2e.feature/feature.xml [hunk 1]
```diff
@@ -2,7 +2,7 @@
 <feature
       id="org.eclipse.m2e.feature"
       label="%featureName"
-      version="1.16.1.qualifier"
+      version="1.16.2.qualifier"
```
Comment (fbricon) [inline org.eclipse.m2e.feature/feature.xml:5]: it's worth bumping the version to 1.17

**Comment by fbricon** [comment]
If you plan on updating the all plugin versions anyways, please do it in a consistent manner, set everything to 1.17.
thx

**Comment by mickaelistria** [comment]
I don't plan to, but I have to because the dependency range were too narrow and update versions had to propagate. I'll update all changed versions to 1.17.

**Comment by mickaelistria** [comment]
Done.

**Comment by mickaelistria** [comment]
@fbricon any progress with impact assessment on m2e-wtp and JBoss Tools?

**Comment by fbricon** [comment]
So m2e-wtp is fine, but JBoss Tools is not. There are 2 calls to MavenPluginActivator.getDefault().getPlexusContainer()
- https://github.com/jbosstools/jbosstools-central/blob/c8b3ebb7876f5383a1fc125fe65ed3e25a6f63cb/maven/plugins/org.jboss.tools.maven.core/src/org/jboss/tools/maven/core/MavenUtil.java#L232
- https://github.com/jbosstools/jbosstools-central/blob/c8b3ebb7876f5383a1fc125fe65ed3e25a6f63cb/maven/plugins/org.jboss.tools.maven.core/src/org/jboss/tools/maven/core/internal/resolution/ArtifactResolutionService.java#L166

**Comment by mickaelistria** [comment]
Thanks @fbricon for checking. I've submitted https://github.com/jbosstools/jbosstools-central/pull/463 to JBoss Tools so we can safely apply the patch to m2e after JBoss Tools does its release.

**Comment by fbricon** [comment]
we can still apply the patch without the removal of the deprecated method (it's an orthogonal issue anyways), without waiting for JBoss Tools to be fixed downstream.

**Comment by mickaelistria** [comment]
As you prefer, it depends on how urgent it is for JDT-LS to consume it. I personally prefer not missing an opportunity to remove deprecated code even if it costs some delay until the merge; but it's something I'm totally flexible with.

**Comment by fbricon** [comment]
I'd rather prefer we merge sooner, even without removing the deprecated methods

**Comment by mickaelistria** [comment]
> I'd rather prefer we merge sooner, even without removing the deprecated methods

OK, I just pushed a new version of the commit on this PR.

File: org.eclipse.m2e.profiles.core/src/org/eclipse/m2e/profiles/core/internal/management/ProfileManager.java [hunk 2]
```diff
@@ -248,9 +248,12 @@ private String findSource(Profile profile, List<Model> modelHierarchy) {
...
       //TODO find an alternative way to get the Maven RepositorySystem, or use Aether directly to resolve models??
-      return MavenPluginActivator.getDefault().getPlexusContainer().lookup(RepositorySystem.class);
-    } catch(ComponentLookupException e) {
-      throw new NoSuchComponentException(e);
+      return MavenPluginActivator.getDefault().getMaven().lookup(RepositorySystem.class);
+    } catch(CoreException e) {
+      if(e.getStatus().getException() instanceof ComponentLookupException) {
+        throw new NoSuchComponentException((ComponentLookupException) e.getStatus().getException());
+      }
+      throw new NoSuchComponentException(null);
```
Comment (fbricon) [inline org.eclipse.m2e.profiles.core/src/org/eclipse/m2e/profiles/core/internal/management/ProfileManager.java:256]: we completely lose the exception information/context that way. At least log the cause of the failure

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java [hunk 3]
```diff
@@ -384,6 +372,24 @@ public MavenRuntimeManagerImpl getMavenRuntimeManager() {
...
+          this.archetypeManager = newArchetypeManager(archetyperContainer, getStateLocation().toFile());
+          try {
+            this.archetypeManager.readCatalogs();
+          } catch(Exception ex) {
+            String msg = "Can't read archetype catalog configuration";
+            log.error(msg, ex);
+          }
+        } catch(PlexusContainerException ex1) {
+          // TODO Auto-generated catch block
+          log.error(ex1.getMessage(), ex1);
```
Comment (fbricon) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java:390]: log.error("Failed to initialize the ArchetypeManager", ex1);

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java [hunk 4]
```diff
@@ -384,6 +372,24 @@ public MavenRuntimeManagerImpl getMavenRuntimeManager() {
...
+          // TODO this is broken, need to make it lazy, otherwise we'll deadlock or timeout... or both 
+          this.archetypeManager = newArchetypeManager(archetyperContainer, getStateLocation().toFile());
+          try {
+            this.archetypeManager.readCatalogs();
+          } catch(Exception ex) {
+            String msg = "Can't read archetype catalog configuration";
+            log.error(msg, ex);
+          }
+        } catch(PlexusContainerException ex1) {
+          // TODO Auto-generated catch block
```
Comment (fbricon) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java:389]: remove TODO

File: org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java [hunk 5]
```diff
@@ -376,6 +348,23 @@ public ProjectRegistryManager getMavenProjectManagerImpl() {
...
+          toDisposeContainers.add(indexerContainer);
+          this.indexManager = new NexusIndexManager(indexerContainer, getMavenProjectManager(), getRepositoryRegistry(),
+              getStateLocation().toFile());
+          getMavenProjectManager().addMavenProjectChangedListener(indexManager);
+          getMaven().addLocalRepositoryListener(new IndexingTransferListener(indexManager));
+          ((RepositoryRegistry) getRepositoryRegistry()).addRepositoryIndexer(indexManager);
+          ((RepositoryRegistry) getRepositoryRegistry())
+              .addRepositoryDiscoverer(new IndexesExtensionReader(indexManager));
+        } catch(PlexusContainerException ex1) {
+          log.error(ex1.getMessage(), ex1);
```
Comment (fbricon) [inline org.eclipse.m2e.core/src/org/eclipse/m2e/core/internal/MavenPluginActivator.java:364]: log.error("Failed to initialize the NexusIndexManager", ex1);

File: pom.xml [hunk 6]
```diff
@@ -34,14 +34,14 @@
 
     <!-- Current target version of m2e, used to deduce some URL at verify and publish.
     This should usually be sync'd with the main feature version -->
-    <m2e.version>1.16.2</m2e.version>
+    <m2e.version>1.17.0</m2e.version>
     <m2e-maven-runtime.version>1.16.0-SNAPSHOT</m2e-maven-runtime.version>
```
Comment (fbricon) [inline pom.xml:38]: 1.17.0-SNAPSHOT

all versions have been bumped elsewhere, so let's stay consistent

**Review by fbricon — CHANGES_REQUESTED** [review]

File: pom.xml (hunk 6 above)
Comment (mickaelistria) [reply pom.xml:38]: Please see comment, this version is used for releng scripts and -SNAPSHOT is not desired in repository URLs (so far).

Comment (fbricon) [reply pom.xml:38]: the comment is for L38, which contains SNAPSHOT already

Comment (mickaelistria) [reply pom.xml:38]: In that case, it's not something we want to do either. If we change that, then it means we have to build newer versions of the maven runtime while it actually didn't change at all, leading to distribution of 10MB of redundant content to the millions of users. It's something we now avoid by decoupling runtime version from consuming m2e version.

**Pull request merged**